ui/src/app/pages/signed-in/inactivity-monitor.tsx

```diff
@@ -0,0 +1,166 @@
+import {Component, Input} from '@angular/core';
+import Timeout = NodeJS.Timeout;
+import {TextModal} from 'app/components/text-modal';
+import {debouncer} from 'app/utils';
+import {ReactWrapperBase} from 'app/utils';
+import {environment} from 'environments/environment';
+
+import * as React from 'react';
+
+const {useState, useEffect} = React;
+
+/*
+ * The user's last known active timestamp is stored in localStorage with the key of
+ * INACTIVITY_CONFIG.LOCAL_STORAGE_KEY_LAST_ACTIVE. This value is checked whenever
+ * the application is reloaded. If the difference between the time at reload and the
+ * value in local storage is greater than the inactivity timeout period, the user will
+ * be signed out of all Google accounts.
+ *
+ * If the localStorage value is null for whatever reason, we defer to the more secure
+ * solution of logging out the user. This should not affect new users since the logout
+ * flow is ignored if there is no user session.
+ */
+export const INACTIVITY_CONFIG = {
+  TRACKED_EVENTS: ['mousedown', 'keypress', 'scroll', 'click'],
+  LOCAL_STORAGE_KEY_LAST_ACTIVE: 'LAST_ACTIVE_TIMESTAMP_EPOCH_MS',
+  MESSAGE_KEY: 'USER_ACTIVITY_DETECTED'
+};
+
+function getInactivityElapsedMs(): number {
+  const lastActive = window.localStorage.getItem(INACTIVITY_CONFIG.LOCAL_STORAGE_KEY_LAST_ACTIVE);
+  if (!lastActive) {
+    return null;
+  }
+  return Date.now() - parseInt(lastActive, 10);
+}
+
+function secondsToText(seconds: number) {
+  return seconds % 60 === 0 && seconds > 60 ?
+      `${seconds / 60} minutes` : `${seconds} seconds`;
+}
+
+export interface Props {
+  signOut: (continuePath?: string) => void;
+}
+
+export const InactivityMonitor = ({signOut}: Props) => {
+  const [showModal, setShowModal] = useState(false);
+
+  function signOutIfLocalStorageInactivityElapsed(continuePath?: string): void {
+    const elapsedMs = getInactivityElapsedMs();
+    if (elapsedMs && elapsedMs > environment.inactivityTimeoutSeconds * 1000) {
+      signOut(continuePath);
+    }
+  }
+
+  // Signal user activity.
+  useEffect(() => {
+    let getUserActivityTimer: () => Timeout;
+    let inactivityInterval: Timeout;
+    let logoutTimer: Timeout;
+    let inactivityModalTimer: Timeout;
+
+    function startUserActivityTracker() {
```

Comment:
nit: this was my laziness in creating the initial version, but most of our code would use `const f = () => {...` style to declare local functions. We should probably have something in our style guide about preference on this as well. There are some subtle differences between these two syntaxes, which are unlikely to matter here, but might matter in other contexts (specifically binding of `this`).

Reply:
Added this to the style guide, and will at least constify the functions I pulled out of the useEffect.